rfc-106-docker-for-local-development.md

```diff
@@ -0,0 +1,37 @@
+# RFC 106: Use Docker for local development
+
+## Summary
+
+Adopt a Docker-based approach for local development instead of the Vagrant VM.
+
+## Problem
+
+We currently rely on the Vagrant development VM for the day to day dev cycle.
+
+- It has [many documented issues](https://docs.publishing.service.gov.uk/manual.html#development-vm)
+- Changes to puppet often result in a broken VM, because the infrastructure team don't use it.
+- We are adopting more cloud-based services into our stack, like S3 and Amazon's Elasticsearch. This means that our development environment doesn't look like production anymore, but we still need code in govuk-puppet.
+- It's hard to keep updated. Running the puppet command will often error and take a long time.
+- GOV.UK is moving to a containerised infrastructure
+
+## Possible approaches to local development
+
+- An approach based on **Docker and docker-compose**. Ben Thorner has a working prototype called [govuk-docker](https://github.com/benthorner/govuk-docker) and we have the [end-to-end tests](https://github.com/alphagov/publishing-e2e-tests) running in the same. Other GDS programmes like GOV.UK Pay use this approach.
+- An approach based on **GNU Guix** package manager, which Chris Baines uses via govuk-guix.
+- Because they are read-only, frontend applications on GOV.UK can be **run against production** using the startup.sh --live flag.
+- Developers on the GOV.UK PaaS work by developing against a **remote set of services** - we’d run many versions of GOV.UK in the cloud, which developers can use in development.
+- Applications can be run locally by **installing dependencies manually**. This is what some of GOV.UK’s Linux users do.
+
+
+Approach | Pro | Cons
+-- | -- | --
+Vagrant VM | Can look like production | Resource intensive, brittle
+Docker | Aligns with future hosting. Highly reproducible. | Slow because on a Mac it needs a VM. Development won’t match production exactly until we migrate Production onto a container-based platform.
+Guix | Huge amount of work put in | Technology unknown in GOV.UK
+Against production | Low setup, fast | Only works for frontend applications. Still needs local install
```

Comment:
I'd add that it also limits you to testing one application at a time. Often a feature will span multiple repositories and it's useful to be able to test it all working together, but the 'against production' approach only lets you test one part of the system at a time and it can be difficult to co-ordinate.